desktop/src/Desktop/ImportExport.hs

```diff
@@ -0,0 +1,234 @@
+
+bipStorageDataKey :: Text
+bipStorageDataKey = "BIPStorage_Data"
+
+storeFrontendVersionKey :: Text
+storeFrontendVersionKey = "StoreFrontend_Version"
+
+storeFrontendDataKey :: Text
+storeFrontendDataKey = "StoreFrontend_Data"
+
+commandLogDataKey :: Text
+commandLogDataKey = "CommandLogs_Data"
+
+commandLogVersionKey :: Text
+commandLogVersionKey  = "CommandLogs_Version"
+
+hoistParser
+  :: Monad m
+  => Text
+  -> StorageVersion
+  -> (Value -> Parser a)
+  -> Value
+  -> ExceptT ImportWalletError m a
+hoistParser errLabel ver p  =
+  hoistEither
+  . first (ImportWalletError_DecodeError errLabel ver . T.pack)
+  . parseEither p
+
+extractImportVersionField
+  :: (Monad m)
+  => Text
+  -> StorageVersion
+  -> Value
+  -> ExceptT ImportWalletError m StorageVersion
+extractImportVersionField key ver =
+  hoistParser key ver (withObject "ChainWeaverImport" (\o -> o .:? key .!= 0 ))
+
+extractImportDataField
+  :: forall a m
+  . ( Monad m
+    , FromJSON a
+    )
+  => Text
+  -> StorageVersion
+  -> Value
+  -> ExceptT ImportWalletError m a
+extractImportDataField key ver =
+  hoistParser key ver (withObject "ChainWeaverImport" (\o -> o .: key))
+
+doImport
+  :: forall t m
+  .  ( MonadIO m
+     , MonadJSM m
+     , HasStorage m
+     , MonadSample t m
+     , Reflex t
+     )
+  => TransactionLogger
+  -> Password -- Password
+  -> Text -- Backup data
+  -> m (Either ImportWalletError (Crypto.XPrv, Password))
+doImport txLogger pw contents = runExceptT $ do
+  jVal <- hoistEither . first (ImportWalletError_NotJson . T.pack) $
+    eitherDecode @Value (TL.encodeUtf8 . TL.fromStrict $ contents)
+
+  bVer <- extractImportVersionField bipStorageVersionKey 0 jVal
+  unless (bVer == 0) $ throwError $ ImportWalletError_UnknownVersion "BIPStorage" bVer
+  bipCrypto <- extractImportDataField @(DMap BIPStorage Identity) bipStorageDataKey 0 jVal
+  rootKey <- failWith ImportWalletError_NoRootKey (runIdentity <$> DMap.lookup BIPStorage_RootKey bipCrypto)
+
+  -- TODO: Test that the password is good for the XPrv.Crypto
+  let
+    testMsg = "test" :: ByteString
+    pwOk = Crypto.verify (Crypto.toXPub rootKey) testMsg $
+       Crypto.sign (T.encodeUtf8 $ unPassword pw) rootKey testMsg
+  unless pwOk $ throwError ImportWalletError_PasswordIncorrect
+
+  feVer <- extractImportVersionField storeFrontendVersionKey 0 jVal
+  feData <- extractImportDataField @Value storeFrontendDataKey feVer jVal
+
+  -- We may not have any commandlogs included in the import so only fail if there is
+  -- something to try to import.
+  _ <- attemptImportCommandLogs jVal
+
+  _ <- ExceptT $ runBIPCryptoT (constant (rootKey, unPassword pw)) $ do
+    let vStore = FrontendStore.versionedStorage
+    feLatestEither <- first (expandDecodeVersionJsonError storeFrontendDataKey feVer)
+      <$> (_versionedStorage_decodeVersionedJson vStore feVer feData)
+
+    traverse
+      (\dmap -> liftIO (TL.putStrLn $ encodeToLazyText dmap) >> _versionedStorage_restoreBackup vStore dmap)
```

Comment:
Because `Show` is for debugging, if we use a specific encoding then we have consistent output rather than raw Haskell data structures that are half pretty printing and half auto-derived.